### PR TITLE
Add support for user defined flags

### PR DESCRIPTION
**Pull request by WanLinghao**
Add support for user defined flags, for example:
apiserver --test-flag=xxx

### CODE REVIEW

File: pkg/cmd/server/start.go [hunk 1]
```diff
@@ -70,16 +70,36 @@ type PostStartHook struct {
...
+	Title            string
+	Version          string
+	TweakConfigFuncs []func(apiServer *apiserver.Config) error
+
+	//FlagConfigFunc handles user-defined flags
+	FlagConfigFuncs []func(*cobra.Command) error
+}
+
 // StartApiServer starts an apiserver hosting the provider apis and openapi definitions.
-func StartApiServer(etcdPath string, apis []*builders.APIGroupBuilder, openapidefs openapi.GetOpenAPIDefinitions, title, version string, tweakConfigFuncs ...func(apiServer *apiserver.Config) error) {
```
Comment (yue9944882) [inline pkg/cmd/server/start.go]: this is breaking backward-compatibility. how about start w/ a new function like `StartApiServerWithOptions` and preserve the old one w/ a deprecation note. and the old one should delegate to the new like:

```go
func StartApiServer(etcdPath string, apis []*builders.APIGroupBuilder, openapidefs openapi.GetOpenAPIDefinitions, title, version string, tweakConfigFuncs ...func(apiServer *apiserver.Config) error) {
  _ = StartApiServerWithOptions(&StartOptions{...})
}
```

File: pkg/cmd/server/start.go [hunk 2]
```diff
@@ -70,16 +70,36 @@ type PostStartHook struct {
...
 
 	signalCh := genericapiserver.SetupSignalHandler()
 	// To disable providers, manually specify the list provided by getKnownProviders()
-	cmd, _ := NewCommandStartServer(etcdPath, os.Stdout, os.Stderr, apis, signalCh, title, version, tweakConfigFuncs...)
+	cmd, _ := NewCommandStartServer(opts.EtcdPath, os.Stdout, os.Stderr, opts.Apis, signalCh,
+		opts.Title, opts.Version, opts.TweakConfigFuncs...)
+
+	for _, ff := range opts.FlagConfigFuncs {
+		if err := ff(cmd); err != nil {
+			panic(err)
```
Comment (yue9944882) [inline pkg/cmd/server/start.go]: how about propagating/returning the error to make the panic opt-out?

Comment (WanLinghao) [reply pkg/cmd/server/start.go]: Do you mean we should return error?

Comment (yue9944882) [reply pkg/cmd/server/start.go]: yea, error

**Comment by WanLinghao** [comment]
@yue9944882  Done PTAL

File: cmd/apiserver-boot/boot/init_repo/repo.go [hunk 3]
```diff
@@ -149,7 +149,12 @@ import (
 
 func main() {
 	version := "v0"
-	server.StartApiServer("/registry/{{ .Domain }}", apis.GetAllApiBuilders(), openapi.GetOpenAPIDefinitions, "Api", version)
+
+	// StartApiServerWithOptions provides more options
+	err := server.StartApiServer("/registry/{{ .Domain }}", apis.GetAllApiBuilders(), openapi.GetOpenAPIDefinitions, "Api", version)
```
Comment (yue9944882) [inline cmd/apiserver-boot/boot/init_repo/repo.go]: lets switch the template to make use of `StartApiServerWithOptions`

**Comment by yue9944882** [comment]
ci looks unhappy

**Comment by WanLinghao** [comment]
@yue9944882  Seems ready to merge

**Pull request merged**